src/main/java/com/o19s/es/ltr/action/TransportAddFeatureToSetAction.java

```diff
@@ -0,0 +1,243 @@
+            }
+            BoolQueryBuilder bq = QueryBuilders.boolQuery();
+            bq.must(nameQuery);
+            bq.must(QueryBuilders.matchQuery("type", StoredFeature.TYPE));
+            srequest.types(IndexFeatureStore.ES_TYPE);
+            srequest.source().query(bq);
+            srequest.source().fetchSource(true);
+            srequest.source().size(StoredFeatureSet.MAX_FEATURES);
+            searchAction.execute(srequest, wrap(this::onSearchResponse, this::onActionFailure));
+
+            GetRequest getRequest = new GetRequest(store)
+                    .type(IndexFeatureStore.ES_TYPE)
+                    .id(StorableElement.generateId(StoredFeatureSet.TYPE, featureSetName))
+                    .routing(routing);
+
+            getRequest.setParentTask(clusterService.localNode().getId(), task.getId());
+            getAction.execute(getRequest, wrap(this::onGetResponse, this::onActionFailure));
+        }
+
+        private void onGetResponse(GetResponse getResponse) {
+            try {
+                StoredFeatureSet featureSet;
+                if (getResponse.isExists()) {
+                    version.set(getResponse.getVersion());
+                    featureSet = IndexFeatureStore.parse(StoredFeatureSet.class, StoredFeatureSet.TYPE, getResponse.getSourceAsBytesRef());
+                } else {
+                    version.set(-1L);
+                    featureSet = new StoredFeatureSet(featureSetName, Collections.emptyList());
+                }
+                setRef.set(featureSet);
+                if (countdown.countDown()) {
+                    mergeAndStore();
+                }
+            } catch (Exception e) {
+                onActionFailure(e);
+            }
+        }
+
+        private void onSearchResponse(SearchResponse sr) {
+            try {
+                if (sr.getHits().getTotalHits() > StoredFeatureSet.MAX_FEATURES) {
+                    throw new IllegalArgumentException("The feature query [" + featureNamesQuery + "] returns too many features");
+                }
+                if (sr.getHits().getTotalHits() == 0) {
+                    throw new IllegalArgumentException("The feature query [" + featureNamesQuery + "] returned no features");
+                }
+                final List<StoredFeature> features = new ArrayList<>(sr.getHits().getHits().length);
+                for (SearchHit hit : sr.getHits().getHits()) {
+                    features.add(IndexFeatureStore.parse(StoredFeature.class, StoredFeature.TYPE, hit.getSourceRef()));
+                }
+                featuresRef.set(features);
+                if (countdown.countDown()) {
+                    mergeAndStore();
+                }
+            } catch (Exception e) {
+                onActionFailure(e);
+            }
+        }
+
+        private void onActionFailure(Exception e) {
+            exception.set(e);
+            if (countdown.countDown()) {
+                listener.onFailure(exception.get());
```

Comment:
Here, exception.get() != e (at least some of the time if both search and get are in error). This is probably a mistake, but I'm not entirely sure of the context.